source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,698 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/io/event_loop.h>
+#include <aws/io/private/event_loop_impl.h>
+
+#include <aws/common/atomics.h>
+#include <aws/common/clock.h>
+#include <aws/common/mutex.h>
+#include <aws/common/rw_lock.h>
+#include <aws/common/task_scheduler.h>
+#include <aws/common/uuid.h>
+
+#include <aws/io/logging.h>
+
+#include <unistd.h>
+
+#include "./dispatch_queue_event_loop_private.h" // private header
+#include <Block.h>
+#include <dispatch/dispatch.h>
+#include <dispatch/queue.h>
+
+static void s_destroy(struct aws_event_loop *event_loop);
+static int s_run(struct aws_event_loop *event_loop);
+static int s_stop(struct aws_event_loop *event_loop);
+static int s_wait_for_stop_completion(struct aws_event_loop *event_loop);
+static void s_schedule_task_now(struct aws_event_loop *event_loop, struct aws_task *task);
+static void s_schedule_task_future(struct aws_event_loop *event_loop, struct aws_task *task, uint64_t run_at_nanos);
+static void s_cancel_task(struct aws_event_loop *event_loop, struct aws_task *task);
+static int s_connect_to_dispatch_queue(struct aws_event_loop *event_loop, struct aws_io_handle *handle);
+static int s_unsubscribe_from_io_events(struct aws_event_loop *event_loop, struct aws_io_handle *handle);
+static void s_free_io_event_resources(void *user_data) {
+    (void)user_data;
+}
+static bool s_is_on_callers_thread(struct aws_event_loop *event_loop);
+
+static struct aws_event_loop_vtable s_vtable = {
+    .destroy = s_destroy,
+    .run = s_run,
+    .stop = s_stop,
+    .wait_for_stop_completion = s_wait_for_stop_completion,
+    .schedule_task_now = s_schedule_task_now,
+    .schedule_task_future = s_schedule_task_future,
+    .cancel_task = s_cancel_task,
+    .connect_to_io_completion_port = s_connect_to_dispatch_queue,
+    .unsubscribe_from_io_events = s_unsubscribe_from_io_events,
+    .free_io_event_resources = s_free_io_event_resources,
+    .is_on_callers_thread = s_is_on_callers_thread,
+};
+
+/**
+ * DISPATCH QUEUE
+ *
+ * Event loop is responsible for processing events and tasks by launching an execution loop on a single thread. Each
+ * iteration of this loop performs three primary jobs:
+ * 1. Process I/O events.
+ * 2. Process cross-thread tasks.
+ * 3. Execute all runnable tasks.
+ *
+ * Apple Dispatch queues are FIFO queues to which the application can submit tasks in the form of block objects, and the
```

Comment:
We should state that we create a serial dispatch queue and not that Apple Dispatch queues are FIFO. Concurrent Apple dispatch queues can be created so we want to be clear here what we are doing.